doc(portal): add javascript asset docs (#1578)
Kong/docs.konghq.com: app/enterprise/1.3-x/developer-portal/theme-customization/single-page-app.md

---

### Introduction

The Kong Developer Portal ships with a default server side rendered theme, however it is possible to replace this with a Single Page App (SPA). Our example is in Angular using the Angular CLI Tool, but you can follow along with any other javascript framework with just a few tweaks.

To view the basic example Angular template from this guide visit the [`example/spa-angular`](https://github.com/Kong/kong-portal-templates/tree/example/spa-angular) branch from the kong-portal-templates branch.

### Prerequisites

* Kong Enterprise 1.3 or later
* Portal Legacy is turned off
* The Developer Portal is enabled and running
* kong-portal-cli tool is installed locally

### What is a SPA

A Single Page App (SPA) is a website that loads all HTML, Javascript, and CSS on the first load. Instead of loading subsequent pages from the server, javascript is used to dynamically change the page content. You may wish to use a SPA in Dev Portal if you have a preexisting SPA you want to integrate with portal, or you are trying to achieve a more application like experience across many pages. A SPA takes control of routing from the server, and handles it client-side instead.

> If you simply wish to add interactive javascript to a few pages, it may make more sense to add the javascript to only layouts that require it, and maintain server side rendering. [Learn more](/enterprise/{{page.kong_version}}/developer-portal/theme-customization/adding-javascript-assets) about adding Javascipt to a layout without implementing a SPA.

Reviewer: ```suggestion
Custom javascript can also be added to run only on specific layouts, allowing you to maintain server-side rendering. [Learn more](/enterprise/{{page.kong_version}}/developer-portal/theme-customization/adding-javascript-assets) about adding javascript to a layout without implementing a SPA.
```

### Making Choices

> We recommend Catalog and Spec routes not be handled by SPA

Reviewer: Can you elaborate on this?

If you are using Authentication, then you probably also want to leave server-side rendering for any account pages

### Getting Started

> Clone down the portal-templates repo

Reviewer: ```suggestion

Clone the [portal-templates](https://github.com/Kong/kong-portal-templates) repo
```

Create a file called `router.conf.yaml` in `workspaces/default` This file will override the default routing, allowing you to control routing via javascript.

`router.conf.yaml` must be a yaml file, where the key is each route, and the value a content or spec path. `/*` Is a catch-all wildcard for all routes not specified in `router.conf.yaml`, it will overwrite all default routing set by collections or set in headmatter.

> In the `router.conf.yaml` example below, we are hardcoding routing for all kong related functionality and deferring to SPA routing for all other routes.  The `/*` route at the bottom of the file is a wildcard route. The wildcard route, in this case, tells Kong to serve `content/index.txt' for any request that is not handled by the route declarations above.

Reviewer: ```suggestion
In the `router.conf.yaml` example below, we are hardcoding routing for all kong related functionality and deferring to SPA routing for all other routes.  The `/*` route at the bottom of the file is a wildcard route. The wildcard route, in this case, tells Kong to serve `content/index.txt` for any request that is not handled by the route declarations above.
```

```
/login: content/login.txt
/logout: content/logout.txt
/register: content/register.txt
/settings: content/settings.txt
/reset-password: content/reset-password.txt
/account/invalidate-verification: content/account/invalidate-verification.txt
/account/resend-verification: content/account/resend-verification.txt
/account/verify: content/account/resend-verification.txt
/documentation: content/documentation/index.txt
/documentation/httpbin: specs/httpbin.json
/documentation/petstore: specs/petstore.yaml
/*: content/index.txt

```

> ### Creating Your SPA

Reviewer: Can you change this section to numbered steps with explicit commands?

Reviewer: Step 1. Creating the SPA
1. 
2.
Step 2. Mounting a SPA

Step 3. Loading your Layout

Step 4. Deploy to Portal

(something like that)

Reviewer 2: @levi I added a numbered list of the steps. Let me know if this is sufficient.

    <script src="assets/js/ng/runtime-es5.js" nomodule defer></script>
    <script src="assets/js/ng/polyfills-es5.js" nomodule defer></script>
    <script src="assets/js/ng/polyfills-es2015.js" type="module"></script>
    <script src="assets/js/ng/styles-es2015.js" type="module"></script>
    <script src="assets/js/ng/styles-es5.js" nomodule defer></script>
    <script src="assets/js/ng/vendor-es2015.js" type="module"></script>
    <script src="assets/js/ng/vendor-es5.js" nomodule defer></script>
    <script src="assets/js/ng/main-es2015.js" type="module"></script>
    <script src="assets/js/ng/main-es5.js" nomodule defer></script>

{-main-}
```
{% endraw %}

If your SPA build process creates a css file edit the head.html partial to include your css file.

### Loading your layout

In order to use our layout modify `workspaces/default/content/index.txt` to use our layout,
The title we set here will be the one that displays until the JS set title loads.

{% raw %}
```
---
layout: spa.html
title: Home
---
```
{% endraw %}

Now using the kong-portal-cli tool we can deploy this portal

From the root folder of the templates repo
```
portal deploy default
```

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

Reviewer: Removes extra newlines as this will fail the linter

---
title: Hosting Single Page App out of Kong Dev Portal
---

### Introduction

> The Kong Developer Portal ships with a default server side rendered theme, however it is possible to replace this with a Single Page App(SPA). Our example is in Angular using the Angular CLI Tool, but you can follow along with any other javascript framework, with just a few tweaks.

Reviewer 3: ```suggestion
The Kong Developer Portal ships with a default server side rendered theme, however it is possible to replace this with a Single Page App (SPA). Our example is in Angular using the Angular CLI Tool, but you can follow along with any other javascript framework with just a few tweaks.
```

> To view the basic example Angular template from this this guide view the [`example/spa-angular` branch from the kong-portal-templates branch](https://github.com/Kong/kong-portal-templates/tree/example/spa-angular).

Reviewer 3: ```suggestion
To view the basic example Angular template from this guide visit the [`example/spa-angular`](https://github.com/Kong/kong-portal-templates/tree/example/spa-angular) branch from the kong-portal-templates branch.
```

### Prerequisites

* Kong Enterprise 1.3 or later
* Portal Legacy is turned off
* The Developer Portal is enabled and running
* kong-portal-cli tool is installed locally

### What is a SPA

> A Single Page App(SPA) is a website that loads all the HTML, Javascript, and CSS on the first load. Instead of loading subsequent pages from the server, javascript is used to dynamically change the page content. You may wish to use a SPA in Dev Portal if you have a preexisting SPA you want to integrate with portal, or you are trying to achieve a more application like experience across many pages. A SPA takes control of routing from the server, and handles it client side instead.

Reviewer 3: ```suggestion
A Single Page App (SPA) is a website that loads all HTML, Javascript, and CSS on the first load. Instead of loading subsequent pages from the server, javascript is used to dynamically change the page content. You may wish to use a SPA in Dev Portal if you have a preexisting SPA you want to integrate with portal, or you are trying to achieve a more application like experience across many pages. A SPA takes control of routing from the server, and handles it client-side instead.
```

> If you simplify wish to add interactive javascript to a few pages, it may make more sense to add the javascript to only layouts that require it, and maintain server side rendering. [Learn more about adding Javascipt to a layout without implementing a SPA.](/enterprise/{{page.kong_version}}/developer-portal/theme-customization/adding-javascript-assets)

Reviewer 3: ```suggestion
If you simply wish to add interactive javascript to a few pages, it may make more sense to add the javascript to only layouts that require it, and maintain server side rendering. [Learn more](/enterprise/{{page.kong_version}}/developer-portal/theme-customization/adding-javascript-assets) about adding Javascipt to a layout without implementing a SPA.
```

### Making Choices

> Decide what pages routing should be handled by SPA and what is handled by content files.

Reviewer 3: I think we can nuke this line
```suggestion
```

We recommend Catalog and Spec routes not be handled by SPA
If you are using Authentication, then you probably also want to leave server-side rendering for any account pages

### Getting Started

> 1. Clone down the portal-templates repo

Reviewer 3: It looks like you started numbering here but then stopped after this section, was that intentional?

2. Create a file called `router.conf.yaml` in `workspaces/default` This file will override the default routing, allowing you to control routing via javascript.

> `router.conf.yaml` must be a yaml file, where the key is each route, and the value a content or spec path. `/*` Is a catch-all wildcard for all routes not specified in `router.conf.yaml`, it will overwrite all default routing from content/spec path or set in headmatter.

Reviewer 3: ```suggestion
`router.conf.yaml` must be a yaml file, where the key is each route, and the value a content or spec path. `/*` Is a catch-all wildcard for all routes not specified in `router.conf.yaml`, it will overwrite all default routing set by collections or set in headmatter.
```

> The following example `router.conf.yaml` will route all paths to the `content/index.txt`, other then routes needed for authentication and spec handling.

Reviewer 3: ```suggestion
In the `router.conf.yaml` example below, we are hardcoding routing for all kong related functionality and deferring to SPA routing for all other routes.  The `/*` route at the bottom of the file is a wildcard route. The wildcard route, in this case, tells Kong to serve `content/index.txt' for any request that is not handled by the route declarations above.
```
